src/checkers/inference/model/serialization/ToStringSerializer.java

```diff
@@ -273,8 +263,15 @@ public String serialize(CombVariableSlot slot) {
     }
 
     protected String indent(String str) {
-        generateIndentations();
-        final StringBuilder sb = new StringBuilder();
+        // create additional indentation strings up to the current level of indentation, if it
+        // doesn't exist in the indentStrings array list
+        // subsequent indentation string values are 1 more INDENT compared to the previous index
+        StringBuilder sb = new StringBuilder(indentStrings.get(indentStrings.size() - 1));
+        for (int i = indentStrings.size(); i <= getIndentationLevel(); i++) {
```

Comment:
StringBuilder is designed to append/deleted frequently, but build to string rarely.

I don't think this will save the performance than just concatenate strings, as now StringBuilder frequently build strings in each iteration. Also, this causes `sb` have to be reused for different usages (temporary string buffers for building caches, and the string builder for building indented string output.)

Reply:
There's no significant runtime bigO advantage, but there is memory allocation savings.

if we used   `indentStrings.add(  indentStrings.get(i-1) + INDENT  );`   the java compiler translates it to

`indentStrings.add(  new StringBuffer(indentStrings.get(i-1)).append(INDENT).toString()  );`

which means every loop iteration we have to reallocate from scratch a brand new block of memory to hold the prior result, and discard the previous stringbuffer instance (which only frees up memory when the garbage collector runs).

As implemented, re keep using one stringbuffer single instance until we have filled the list with the required number of strings.

Reply:
That is over-engineering...

we will not indent too much levels, maybe 4~5 would be the most in practice. Personally, I think make code easier to understand (i.e. do not reuse `sb` for different usages) takes over than this "memory savings"... But if you insist, I'm fine with this.

Reply:
I remember prof said last week that I should reuse the same `sb` instance instead of doing `indentStrings.get(i-1) + INDENT` in a loop. ¯\_(ツ)_/¯

Reply:
For this question, I think reusing the same `sb` in the `for` block is ok, as it is still used to hold the temporary string to fill missing strings in the list. The latter one(line 274) reuses the same variable `sb`, but the instance isn't reusing the previous one(newly created instead). I'm fine with this.

Reply:
I'll make it cleaner by using a fresh variable name for the latter instance.

Reply:
We were talking about over-engineering and how you could also use StringBuffers.
I agree that readable code is much more important than this micro-optimization. We don't indent much and this code isn't used on the critical path.
Make the code clear and maintainable and let's optimize the critical path.